scripts/build/builders/builder.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+
+import logging
+import os
+import shutil
+from abc import ABC, abstractmethod
+
+from shellrunner import ShellRunner
+
+
+class Builder(ABC):
+  """Generic builder base class for CHIP.
+
+  Provides ability to boostrap and copy output artifacts and subclasses can use
+  a generic shell runner.
+
+  """
+
+  def __init__(self, root, output_dir='out'):
+    self.root = os.path.abspath(root)
+    self._runner = ShellRunner()
+    self.output_dir = output_dir
+
+  @abstractmethod
+  def generate(self):
+    """Generate the build files - generally the ninja/makefiles"""
+    raise NotImplementedError()
+
+  @abstractmethod
+  def build(self):
+    """Perform an actual build"""
+    raise NotImplementedError()
+
+  @abstractmethod
+  def outputs(self):
+    """Return a list of relevant output files after a build.
+
+       May use build output data (e.g. manifests), so this should be invoked
+       only after a build has succeeded.
+    """
+    raise NotImplementedError()
+
+  def _Execute(self, cmdarray, **args):
+    self._runner.Run(cmdarray, **args)
+
+  def _ActivatedExecute(self, cmdstring: str, **args):
+    """Switches to top level to activate then runs the given command in the original run directory."""
+    self._runner.Run([
+        'bash', '-c',
+        'cd "%s"; source ./scripts/activate.sh; cd -; %s' %
+        (self.root, cmdstring)
+    ], **args)
+
+  def Bootstrap(self):
+    envpath = os.environ.get('PW_ENVIRONMENT_ROOT',
+                             os.path.join(self.root, '.environment'))
+
+    if not os.path.exists(os.path.join(envpath, 'activate.sh')):
```

Comment:
What's this for?

Reply:
Runs 'bootstrap.sh' if not already run (causes the pigweed binaries to be downloaded for the platform).
I did not want to check for directory existence, so I checked for a file that seems reasonably expected to be there instead.

Reply:
activate.sh already does that

Reply:
Will work towards assuming 'bootstrap/activate was already called'